### DISCUSSION TITLE
Validation inside Form component - required_if don't work.

### DISCUSSION THREAD

**Discussion by peroxovy** [livewire/livewire#6280] [1. Help]
Hello, I'm using Livewire 3 for tests and it seems that if I put inside the `FormComponent` Rules like:

```php
    #[Rule('required|string|exists:types,type_key')]
    public $type = 'text';

    #[Rule('required_if:type,text|string')]
    public $text;

    #[Rule('required_if:type,name|string')]
    public $name;
```
It seems  that it does not work properly. When I `dd($this->form);` inside the `Post` Livewire component I see all fields from form...

Currently I have `<select>` inside form, which is responsible for showing/hiding fields in it depending on `<select wire:model="type">` value:

` <x-forms.select-option :options="$types" wire:model.live="type" required/>`

Implementation how I show fields inside Component:

```html
@if($form->type == 'text')
<div class="p-2 block">
      <x-forms.label-text for="form.text">Text</x-forms.label-text>
      <x-forms.input-text wire:model.defer="form.text"/>
      <div>@error('form.text') {{ $message }} @enderror</div>
</div>
@endif
```
The inputs with rule value `required_if:` shows an error like : `The text field must be a string.` when I switch, to the `type` that shouldn't be taken into consideration.

This is how I done checking:
```php
class Post extends Component
{
    public FormComponent $form;

    public function render()
    {
        $types = app(GetAllTypesAsKeyValueArrayService::class)->execute();

        return view('livewire.post', ['types' => $types]);
    }

    public function store()
    {
        dd($this->form);
    }
```

F.e. I want to see 2 values passed to component, but I see all of them (let's say 5), with `null` value (filled inputs are okay with values).

In past Livewire versions, when I initialize `$rules = [ ]` array in single all things was working. Now with new version, when I want to separate the `form` logic and validation, problems come.

**Comment by peroxovy** [comment]
No ideas till 3 weeks?

**Comment by uadoronin** [comment]
Same problem in Form Component:

```
    protected function rules(): array
        {
            return [
                'email' => 'required|email',
                'code' => 'string|required_with:email'
        ];
    }
```

Conditional validation only works when we use form prefix:
`'code' => 'string|required_with:form.email'`

**Comment by madushancs** [comment]
any solution?

**Reply by uadoronin** [reply]
I think the solution is to overwrite functionality in Livewire\Features\SupportFormObjects\Form
We need to change `$value` property. 

```
    protected function getAttributesWithPrefixedKeys($attributes)
    {
        $attributesWithPrefixedKeys = [];

        foreach ($attributes as $key => $value) {
            $attributesWithPrefixedKeys[$this->propertyName . '.' . $key] = $value;
        }

        return $attributesWithPrefixedKeys;
    }
```

But it is problematic.
It must support validation (with fields) as:
```
[accepted_if:anotherfield,value,...]
[declined_if:anotherfield,value,...]
[different:field]
[exclude_if:anotherfield,value]
[exclude_unless:anotherfield,value]
[exclude_with:anotherfield]
[exclude_without:anotherfield]
[gt:field]
[gte:field]
[in_array:anotherfield.*]
[lt:field]
[lte:field]
[missing_if:anotherfield,value,...]
[missing_unless:anotherfield,value]
[missing_with:foo,bar,...]
[missing_with_all:foo,bar,...]
[prohibited_if:anotherfield,value,...]
[prohibited_unless:anotherfield,value,...]
[prohibits:anotherfield,...]
[required_if:anotherfield,value,...]
[required_unless:anotherfield,value,...]
[required_with:foo,bar,...]
[required_with_all:foo,bar,...]
[required_without:foo,bar,...]
[required_without_all:foo,bar,...]
```
Then we can use (for example):
```
    #[Rule('required_if:anotherfield')]
    public $dob;
```
And Form Component auto generate validation rule as:
`"form.dob" => "required_if:form.anotherfield"`

**Comment by nicolagianelli** [comment]
#6614 This PR should fix this, if you want to test it while Caleb decide if it's ok to merge it

**Reply by peroxovy** [reply]
Thanks for your work. I'll test it asap :)

**Comment by lewislarsen** [comment]
Still encountering this issue in Livewire 3.5.0. I have a form component and my test is failing any `required_if` rule that is set in the extracted component. It would be nice to see a resolution to this problem.

**Reply by nicolagianelli** [reply]
Hello, my PR wasn't merged so the problem still persist.

**Reply by ndarproj** [reply]
same issue :(

**Comment by lcallif** [comment]
Have you tried to create a "public function messages()" with the rules in the function executing the validation.  Please refer to the attached screen shots for what I am referring to.  Also, please note that I am NOT using a "required_if" rule.
If the above approach works then you can narrow down your Issue report for the Livewire folks.

``    public function addPayment($id)
    {
        $validated = $this->validate([
            'paymentDt' => 'required',
            'paymenttype' => 'required|gt:0',
            'paymentamount' => 'required|gt:0',
            'paymentdescription' => 'required|min:10',
        ]);
        CustomerProposalProjectPayment::create([
                'user_id' => Auth::id(),
                'team_id' => Auth::user()->currentTeam->id,
                'customerproposalproject_id' => $id,
                'typeId' => $this->paymenttype,
                'paymentDt' => $this->paymentDt,
                'paymentAmount' => $this->paymentamount,
                'description' => $this->paymentdescription,
        ]);
        return redirect($this->redirectUrl);
    }

    public function messages()
    {
        return [
            'paymenttype.required' => 'Payment Type is Required.',
            'paymentdescription.required' => 'Description is a minimum of 10 characters.',
            'paymentamount.required' => 'Payment Amount is Required.',
            'paymentDt.required' => 'Payment Date is Required.',
        ];
    }
``
